### PR TITLE
change constraint for foldMapM from Monad to Applicative

### PR DESCRIPTION
**Pull request by mberndt123**
Applicative is enough for this method, so it shouldn't require Monad.

### CODE REVIEW

**Comment by travisbrown** [comment]
This could probably be done in a binary-compatible way by making the current implementation package private and adding the new one as an overload. I agree with @LukaJCB that it probably makes more sense to add a new `foldMapA` with the appropriate constraint instead.

It seems like the bigger issue with this implementation is short circuiting. For example, this will hang forever with the `foldMap` implementation:

```scala
import cats.implicits._

Stream.iterate(0)(_ + 1).foldMapM(b => if (b > 10) Left(b) else Right(b))
```
I think an implementation like this should be fine both in terms of stack safety and short-circuiting:

```scala
def foldMapA[G[_], A, B](fa: F[A])(f: A => G[B])(implicit G: Applicative[G], B: Monoid[B]): G[B] =
  foldRight(fa, Eval.now(G.pure(B.empty)))((a, egb) =>
    G.map2Eval(f(a), egb)(B.combine)
  ).value
```

For example:

```scala
scala> import cats.implicits._
import cats.implicits._

scala> Stream.iterate(0)(_ + 1).foldMapA(b => if (b > 10) Left(b) else Right(b))
res0: scala.util.Either[Int,Int] = Left(11)

scala> Stream.iterate(0)(_ + 1).foldMapA(b => if (b > 1000000) Left(b) else Right(b))
res1: scala.util.Either[Int,Int] = Left(1000001)
```

**Comment by mberndt123** [comment]
Hey there,

Thanks for the review! I didn't know about map2Eval, that's very interesting.

I've now added `foldMapA` as a new method and also a test to make sure the short-circuiting behaviour is preserved.

File: core/src/main/scala/cats/Foldable.scala [hunk 1]
```diff
@@ -333,6 +333,14 @@ import Foldable.sentinel
   def foldMapM[G[_], A, B](fa: F[A])(f: A => G[B])(implicit G: Monad[G], B: Monoid[B]): G[B] =
     foldM(fa, B.empty)((b, a) => G.map(f(a))(B.combine(b, _)))
 
+  /**
+   * Equivalent to foldMapM.
+   * The difference is that foldMapA only requires G to be an Applicative
+   * rather than a Monad. It is also slower due to use of Eval.
```
Comment (travisbrown) [inline core/src/main/scala/cats/Foldable.scala]: Do we actually know that this is true? My intuition is that it generally will be, but there's also some `Either` overhead involved in the `tailRecM` implementation.

**Comment by travisbrown** [comment]
I put together a little benchmark with a vector with 10k elements with examples that both short-circuit in `Either` and complete successfully, and yeah, the `foldMapM` implementation is a lot faster:

```
[info] Benchmark                            Mode  Cnt     Score     Error  Units
[info] FoldMapABench.foldMapA              thrpt   20  2113.567 ± 137.868  ops/s
[info] FoldMapABench.foldMapM              thrpt   20  3445.172 ± 131.904  ops/s
[info] FoldMapABench.foldMapAShortCircuit  thrpt   20  5476.648 ± 289.969  ops/s
[info] FoldMapABench.foldMapMShortCircuit  thrpt   20  6699.230 ± 259.521  ops/s
```

**Comment by LukaJCB** [comment]
Let's update the scaladoc and get this merged :)

**Comment by mberndt123** [comment]
@LukaJCB I'm not quite sure what more I'm supposed to add for scaladoc. Isn't it enough to say that this does the same thing as `foldMapM`?

**Comment by LukaJCB** [comment]
Sorry had a brainfart there :smile:

**Pull request merged**